alacritty_terminal/src/grid/storage.rs

```diff
@@ -173,27 +176,29 @@ impl<T> Storage<T> {
         }
     }
 
-    /// Truncate the invisible elements from the raw buffer
+    /// Truncate the invisible elements from the raw buffer.
     pub fn truncate(&mut self) {
         self.inner.rotate_left(self.zero);
         self.inner.truncate(self.len);
 
         self.zero = 0;
     }
 
-    /// Dynamically grow the storage buffer at runtime
-    pub fn initialize(&mut self, num_rows: usize, template_row: Row<T>)
+    /// Dynamically grow the storage buffer at runtime.
+    pub fn initialize(&mut self, additional: usize, template_row: Row<T>)
     where
         T: Clone,
     {
-        let mut new = vec![template_row; num_rows];
-
-        let mut split = self.inner.split_off(self.zero);
-        self.inner.append(&mut new);
-        self.inner.append(&mut split);
+        if self.len + additional > self.inner.len() {
```

Comment:
Adding a comment above this `if` is probably a good idea. Something like this:

```suggestion
        // Skip initializing new rows if enough are already buffered
        if self.len + additional > self.inner.len() {
```